File: vehicles/excavator/articulation/ExcavatorModel.cpp
#include "vehicles/excavator/articulation/ExcavatorModel.h"

#include <cmath>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unordered_map>

#include "chrono/assets/ChVisualShapeBox.h"
#include "chrono/assets/ChVisualShapeSphere.h"
#include "chrono/assets/ChVisualShapeTriangleMesh.h"
#include "chrono/functions/ChFunctionConst.h"
#include "chrono/geometry/ChTriangleMeshConnected.h"
#include "chrono/collision/ChCollisionShapeBox.h"
#include "chrono/physics/ChBodyEasy.h"
#include "chrono/physics/ChContactMaterialNSC.h"
#include "chrono/physics/ChLinkMotionImposed.h"
#include "chrono/physics/ChLinkMotorRotationAngle.h"
#include "chrono/physics/ChLinkMotorRotationSpeed.h"
#include "chrono/physics/ChSystem.h"
#include "vehicles/common/DifferentialDriveFunctions.h"

namespace aes {
namespace {

using chrono::ChBody;
using chrono::ChBodyEasyBox;
using chrono::ChColor;
using chrono::ChFrame;
using chrono::ChTriangleMeshConnected;
using chrono::ChVector3d;

constexpr double kMathScavatorScale = 2.8 / 6.24;

struct MeshCalibration {
    ChVector3d translation;
    ChVector3d rotation_degrees;
    double scale = 1.0;
};

std::string AssetPath(const std::string& filename) {
    std::string path = AES_EXCAVATOR_ASSET_DIR;
    if (!path.empty() && path.back() != '/' && path.back() != '\\')
        path += '/';
    return path + filename;
}

std::unordered_map<std::string, MeshCalibration> LoadVisualCalibration(const std::string& path) {
    std::ifstream input(path);
    if (!input)
        throw std::runtime_error("Could not open excavator visual calibration file: " + path);

    std::unordered_map<std::string, MeshCalibration> result;
    std::string line;
    std::size_t line_number = 0;
    while (std::getline(input, line)) {
        ++line_number;
        if (const auto comment = line.find('#'); comment != std::string::npos)
            line.erase(comment);

        std::istringstream values(line);
        std::string part;
        MeshCalibration calibration;
        if (!(values >> part))
            continue;
        if (!(values >> calibration.translation.x() >> calibration.translation.y() >> calibration.translation.z() >>
              calibration.rotation_degrees.x() >> calibration.rotation_degrees.y() >>
              calibration.rotation_degrees.z() >> calibration.scale)) {
            throw std::runtime_error("Invalid excavator visual calibration at " + path + ":" +
                                     std::to_string(line_number));
        }
        std::string unexpected;
        if (values >> unexpected)
            throw std::runtime_error("Unexpected value in excavator visual calibration at " + path + ":" +
                                     std::to_string(line_number));
        if (part != "base" && part != "boom" && part != "arm" && part != "bucket")
            throw std::runtime_error("Unknown excavator part '" + part + "' in " + path + ":" +
                                     std::to_string(line_number));
        if (!std::isfinite(calibration.scale) || calibration.scale <= 0.0)
            throw std::runtime_error("Excavator visual scale must be positive at " + path + ":" +
                                     std::to_string(line_number));
        if (!result.emplace(part, calibration).second)
            throw std::runtime_error("Duplicate excavator part '" + part + "' in " + path);
    }

    for (const std::string part : {"base", "boom", "arm", "bucket"}) {
        if (result.count(part) == 0)
            throw std::runtime_error("Missing excavator part '" + part + "' in " + path);
    }
    return result;
}

ChFrame<> CalibrationFrame(const MeshCalibration& calibration) {
    const auto radians = calibration.rotation_degrees * (chrono::CH_PI / 180.0);
    const auto rotation = chrono::QuatFromAngleZ(radians.z()) * chrono::QuatFromAngleY(radians.y()) *
                          chrono::QuatFromAngleX(radians.x());
    return {calibration.translation, rotation};
}

std::shared_ptr<ChBody> Box(chrono::ChSystem& system,
                            const ChVector3d& size,
                            const ChVector3d& position,
                            const ChColor& color,
                            bool visible = true,
                            bool collidable = false) {
    std::shared_ptr<chrono::ChContactMaterialNSC> material;
    if (collidable) {
        material = chrono_types::make_shared<chrono::ChContactMaterialNSC>();
        material->SetFriction(0.8f);
    }
    auto body = chrono_types::make_shared<ChBodyEasyBox>(size.x(), size.y(), size.z(), 780.0, visible,
                                                         collidable, material);
    body->SetPos(position);
    if (visible)
        body->GetVisualShape(0)->SetColor(color);
    system.Add(body);
    return body;
}

void AddBoxVisual(const std::shared_ptr<ChBody>& body,
                  const ChVector3d& size,
                  const ChVector3d& position,
                  const ChColor& color) {
    auto shape = chrono_types::make_shared<chrono::ChVisualShapeBox>(size.x(), size.y(), size.z());
    shape->SetColor(color);
    body->AddVisualShape(shape, ChFrame<>(position));
}

void StyleExcavatorCab(const std::shared_ptr<ChBody>& cabin) {
    const ChColor glass{0.08f, 0.18f, 0.22f};
    AddBoxVisual(cabin, {0.035, 1.35, 0.72}, {1.01, 0.0, 0.18}, glass);
    AddBoxVisual(cabin, {1.20, 0.035, 0.72}, {0.20, 0.91, 0.18}, glass);
    AddBoxVisual(cabin, {1.20, 0.035, 0.72}, {0.20, -0.91, 0.18}, glass);
    AddBoxVisual(cabin, {2.15, 1.95, 0.12}, {0.0, 0.0, 0.80}, {0.92f, 0.58f, 0.02f});
}

void StyleExcavatorTrack(const std::shared_ptr<ChBody>& chassis, const ChVector3d& center) {
    // Add inexpensive individual grouser plates so the tracks read as tracks
    // even at presentation distance, while retaining a simple collision box.
    AddBoxVisual(chassis, {3.8, 0.55, 0.65}, center, {0.12f, 0.12f, 0.12f});
    constexpr int shoes = 13;
    for (int index = 0; index < shoes; ++index) {
        const double x = -1.72 + 3.44 * index / (shoes - 1);
        AddBoxVisual(chassis, {0.20, 0.60, 0.07}, center + ChVector3d{x, 0.0, -0.36},
                     {0.20f, 0.20f, 0.19f});
        AddBoxVisual(chassis, {0.20, 0.60, 0.07}, center + ChVector3d{x, 0.0, 0.36},
                     {0.20f, 0.20f, 0.19f});
    }
}

void AddTrackCollision(const std::shared_ptr<ChBody>& chassis, const ChVector3d& center) {
    auto material = chrono_types::make_shared<chrono::ChContactMaterialNSC>();
    material->SetFriction(0.8f);
    chassis->AddCollisionShape(
        chrono_types::make_shared<chrono::ChCollisionShapeBox>(material, 3.8, 0.55, 0.72), ChFrame<>(center));
    chassis->EnableCollision(true);
}

void StyleTrackRotor(const std::shared_ptr<ChBody>& rotor) {
    AddBoxVisual(rotor, {0.45, 0.50, 0.45}, {0, 0, 0}, {0.18f, 0.18f, 0.17f});
    AddBoxVisual(rotor, {0.12, 0.58, 0.10}, {0, 0, 0.25}, {0.95f, 0.65f, 0.05f});
}

std::shared_ptr<ChTriangleMeshConnected> LoadMathScavatorMesh(const std::string& filename, double scale) {
    const auto path = AssetPath(filename);
    const auto mesh = ChTriangleMeshConnected::CreateFromSTLFile(path);
    if (!mesh || mesh->GetNumVertices() == 0)
        throw std::runtime_error("Could not load MathScavator9000 vehicle asset: " + path);
    // Chrono 10's Irrlicht triangle-mesh path copies raw vertices and does not
    // apply ChVisualShapeTriangleMesh::SetScale().  Bake the user calibration
    // scale into this private mesh instance so every renderer sees it.
    mesh->Transform(ChVector3d{0, 0, 0}, chrono::ChMatrix33d(scale));
    return mesh;
}

void SetImportedVisual(const std::shared_ptr<ChBody>& body,
                       const std::shared_ptr<ChTriangleMeshConnected>& mesh,
                       const std::string& name,
                       const ChFrame<>& frame,
                       const ChColor& color) {
    body->GetVisualModel()->Clear();
    auto shape = chrono_types::make_shared<chrono::ChVisualShapeTriangleMesh>(mesh, false);
    shape->SetName(name);
    shape->SetColor(color);
    body->AddVisualShape(shape, frame);
}

void AddJointMarker(const std::shared_ptr<ChBody>& body, const ChColor& color) {
    auto marker = chrono_types::make_shared<chrono::ChVisualShapeSphere>(0.16);
    marker->SetColor(color);
    // Put the marker just outside the near face of the link so the imported
    // mesh cannot hide it.  The actual hinge centre is directly behind it.
    body->AddVisualShape(marker, ChFrame<>(ChVector3d{0.0, -0.42, 0.0}));
}

void AddBucketCollision(const std::shared_ptr<ChBody>& bucket) {
    auto material = chrono_types::make_shared<chrono::ChContactMaterialNSC>();
    material->SetFriction(0.8f);
    // Bounding box of stick_bucket_link.STL after the common source-to-simulator rotation and scale.
    bucket->AddCollisionShape(
        chrono_types::make_shared<chrono::ChCollisionShapeBox>(material, 0.88, 0.52, 1.00),
        ChFrame<>(ChVector3d{-0.384, -0.002, 0.044}));
    bucket->EnableCollision(true);
}

std::shared_ptr<chrono::ChLinkMotorRotationAngle> AngleMotor(
    chrono::ChSystem& system,
    const std::shared_ptr<ChBody>& child,
    const std::shared_ptr<ChBody>& parent,
    const ChVector3d& pivot,
    const chrono::ChQuaternion<>& orientation = chrono::QUNIT) {
    auto motor = chrono_types::make_shared<chrono::ChLinkMotorRotationAngle>();
    motor->Initialize(child, parent, ChFrame<>(pivot, orientation));
    motor->SetAngleFunction(chrono_types::make_shared<chrono::ChFunctionConst>(0.0));
    system.Add(motor);
    return motor;
}

}  // namespace

ExcavatorModel::ExcavatorModel(chrono::ChSystem& system,
                               bool show_rigid_ground,
                               bool show_calibration_markers)
    : system_(system), visual_calibration_path_(AssetPath("visual_calibration.cfg")) {
    const auto initial_body_count = system_.GetBodies().size();
    const auto calibration = LoadVisualCalibration(visual_calibration_path_);
    ground_ = Box(system_, {30, 20, 0.1}, {4, 0, -0.05}, {0.38f, 0.30f, 0.18f}, show_rigid_ground);
    ground_->SetFixed(true);

    chassis_ = Box(system_, {3.6, 2.4, 0.45}, {0, 0, 0.75}, {0.22f, 0.22f, 0.20f});
    cabin_ = Box(system_, {2.0, 1.8, 1.5}, {0, 0, 1.75}, {0.95f, 0.65f, 0.05f});
    // Small visible rotors prove independent track commands without rotating the
    // entire track loop as if it were a rigid wheel.
    left_track_ = Box(system_, {0.20, 0.20, 0.20}, {0, 1.35, 0.48}, {0.12f, 0.12f, 0.12f}, false);
    right_track_ = Box(system_, {0.20, 0.20, 0.20}, {0, -1.35, 0.48}, {0.12f, 0.12f, 0.12f}, false);
    StyleTrackRotor(left_track_);
    StyleTrackRotor(right_track_);
    // Use the scaled upstream URDF link origins as the body frames.  This makes
    // every imported mesh origin coincide with its actual Chrono hinge instead
    // of requiring unrelated per-part guesses.
    const ChVector3d boom_pivot{0.75, 0, 2.25};
    const ChVector3d arm_pivot = boom_pivot + ChVector3d{6.24 * kMathScavatorScale, 0, 0};
    const ChVector3d bucket_pivot =
        arm_pivot + ChVector3d{-0.0091425 * kMathScavatorScale, 0, -2.9796 * kMathScavatorScale};
    boom_ = Box(system_, {2.8, 0.35, 0.35}, boom_pivot, {0.95f, 0.65f, 0.05f});
    arm_ = Box(system_, {2.2, 0.30, 0.30}, arm_pivot, {0.95f, 0.65f, 0.05f});
    bucket_ = Box(system_, {0.9, 0.52, 1.0}, bucket_pivot, {0.90f, 0.55f, 0.03f});
    StyleExcavatorCab(cabin_);

    const auto& base_calibration = calibration.at("base");
    const auto& boom_calibration = calibration.at("boom");
    const auto& arm_calibration = calibration.at("arm");
    const auto& bucket_calibration = calibration.at("bucket");
    const auto base_mesh = LoadMathScavatorMesh("base_link.STL", base_calibration.scale);
    const auto boom_mesh = LoadMathScavatorMesh("chassis_boom_link.STL", boom_calibration.scale);
    const auto arm_mesh = LoadMathScavatorMesh("boom_stick_link.STL", arm_calibration.scale);
    const auto bucket_mesh = LoadMathScavatorMesh("stick_bucket_link.STL", bucket_calibration.scale);
    imported_vertex_count_ = base_mesh->GetNumVertices() + boom_mesh->GetNumVertices() +
                             arm_mesh->GetNumVertices() + bucket_mesh->GetNumVertices();

    SetImportedVisual(chassis_, base_mesh, "MathScavator9000 undercarriage",
                      CalibrationFrame(base_calibration),
                      {0.20f, 0.20f, 0.20f});
    SetImportedVisual(boom_, boom_mesh, "MathScavator9000 boom",
                      CalibrationFrame(boom_calibration),
                      {0.95f, 0.65f, 0.05f});
    SetImportedVisual(arm_, arm_mesh, "MathScavator9000 stick",
                      CalibrationFrame(arm_calibration),
                      {0.95f, 0.65f, 0.05f});
    SetImportedVisual(bucket_, bucket_mesh, "MathScavator9000 bucket",
                      CalibrationFrame(bucket_calibration),
                      {0.90f, 0.55f, 0.03f});
    AddBucketCollision(bucket_);
    if (show_calibration_markers) {
        AddJointMarker(boom_, {1.0f, 0.0f, 0.8f});
        AddJointMarker(arm_, {0.1f, 1.0f, 0.1f});
        AddJointMarker(bucket_, {0.0f, 0.9f, 1.0f});
    }
    const ChVector3d left_track_center{0.0, 1.35, -0.27};
    const ChVector3d right_track_center{0.0, -1.35, -0.27};
    StyleExcavatorTrack(chassis_, left_track_center);
    StyleExcavatorTrack(chassis_, right_track_center);
    AddTrackCollision(chassis_, left_track_center);
    AddTrackCollision(chassis_, right_track_center);

    // Impose continuous differential-drive motion from separate left/right track commands.
    drive_state_ = std::make_shared<detail::DifferentialDriveState>(2.7);
    drive_ = chrono_types::make_shared<chrono::ChLinkMotionImposed>();
    drive_->Initialize(chassis_, ground_, ChFrame<>(chassis_->GetPos()));
    drive_->SetPositionFunction(
        chrono_types::make_shared<detail::DifferentialDrivePositionFunction>(drive_state_));
    drive_->SetRotationFunction(
        chrono_types::make_shared<detail::DifferentialDriveRotationFunction>(drive_state_));
    system_.Add(drive_);

    // Swing uses world Z; implement attachment joints as motors so every required DOF is commandable.
    joints_[0] = AngleMotor(system_, cabin_, chassis_, {0, 0, 1.0});
    const auto hinge_y = chrono::QuatFromAngleX(chrono::CH_PI_2);
    joints_[1] = AngleMotor(system_, boom_, cabin_, boom_pivot, hinge_y);
    joints_[2] = AngleMotor(system_, arm_, boom_, arm_pivot, hinge_y);
    joints_[3] = AngleMotor(system_, bucket_, arm_, bucket_pivot, hinge_y);

    left_track_drive_ = chrono_types::make_shared<chrono::ChLinkMotorRotationSpeed>();
    left_track_drive_->Initialize(left_track_, chassis_, ChFrame<>({0, 1.35, 0.48}, hinge_y));
    left_track_drive_->SetSpeedFunction(chrono_types::make_shared<chrono::ChFunctionConst>(0.0));
    system_.Add(left_track_drive_);

    right_track_drive_ = chrono_types::make_shared<chrono::ChLinkMotorRotationSpeed>();
    right_track_drive_->Initialize(right_track_, chassis_, ChFrame<>({0, -1.35, 0.48}, hinge_y));
    right_track_drive_->SetSpeedFunction(chrono_types::make_shared<chrono::ChFunctionConst>(0.0));
    system_.Add(right_track_drive_);

    body_count_ = system_.GetBodies().size() - initial_body_count;
}

void ExcavatorModel::SetDriveSpeeds(double left, double right) {
    drive_state_->SetSpeeds(system_.GetChTime(), left, right);
    left_track_drive_->SetSpeedFunction(chrono_types::make_shared<chrono::ChFunctionConst>(left));
    right_track_drive_->SetSpeedFunction(chrono_types::make_shared<chrono::ChFunctionConst>(right));
}

void ExcavatorModel::SetArticulationTargets(double swing, double boom, double arm, double bucket) {
    const std::array<double, 4> targets{swing, boom, arm, bucket};
    for (std::size_t i = 0; i < joints_.size(); ++i)
        joints_[i]->SetAngleFunction(chrono_types::make_shared<chrono::ChFunctionConst>(targets[i]));
}

chrono::ChVector3d ExcavatorModel::GetChassisPosition() const { return chassis_->GetPos(); }

double ExcavatorModel::GetChassisHeading() const { return chassis_->GetRot().GetCardanAnglesXYZ().z(); }

std::array<double, 4> ExcavatorModel::GetJointAngles() const {
    return {joints_[0]->GetMotorAngle(), joints_[1]->GetMotorAngle(), joints_[2]->GetMotorAngle(),
            joints_[3]->GetMotorAngle()};
}

std::array<double, 2> ExcavatorModel::GetTrackAngles() const {
    return {left_track_drive_->GetMotorAngle(), right_track_drive_->GetMotorAngle()};
}

std::array<chrono::ChVector3d, 3> ExcavatorModel::GetArticulationBodyPositions() const {
    return {boom_->GetPos(), arm_->GetPos(), bucket_->GetPos()};
}

std::shared_ptr<chrono::ChBody> ExcavatorModel::GetLeftTrackBody() const { return chassis_; }

std::shared_ptr<chrono::ChBody> ExcavatorModel::GetRightTrackBody() const { return chassis_; }

std::shared_ptr<chrono::ChBody> ExcavatorModel::GetBucketBody() const { return bucket_; }

std::size_t ExcavatorModel::GetBodyCount() const { return body_count_; }

std::size_t ExcavatorModel::GetImportedVertexCount() const { return imported_vertex_count_; }

const std::string& ExcavatorModel::GetVisualCalibrationPath() const { return visual_calibration_path_; }

}  // namespace aes

File: vehicles/excavator/articulation/excavator_model.py
"""MathScavator9000 excavator model implemented with PyChrono."""

from __future__ import annotations

from dataclasses import dataclass
import math
from pathlib import Path

import pychrono as chrono

from vehicles.common.chrono_utils import (
    add_box_collision,
    add_box_visual,
    color,
    make_angle_motor,
    make_box,
    vector,
)
from vehicles.common.differential_drive import DifferentialDrive


ASSET_DIR = Path(__file__).resolve().parents[1] / "model" / "assets" / "mathscavator9000"
MATHSCAVATOR_SCALE = 2.8 / 6.24
YELLOW = (0.95, 0.65, 0.05)


@dataclass(frozen=True)
class MeshCalibration:
    translation: tuple[float, float, float]
    rotation_degrees: tuple[float, float, float]
    scale: float


def _load_calibration(path: Path) -> dict[str, MeshCalibration]:
    result: dict[str, MeshCalibration] = {}
    for line_number, raw_line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        values = raw_line.split("#", 1)[0].split()
        if not values:
            continue
        if len(values) != 8:
            raise ValueError(f"Invalid excavator visual calibration at {path}:{line_number}")
        part = values[0]
        if part not in {"base", "boom", "arm", "bucket"}:
            raise ValueError(f"Unknown excavator part {part!r} at {path}:{line_number}")
        numbers = tuple(float(value) for value in values[1:])
        if not math.isfinite(numbers[6]) or numbers[6] <= 0.0:
            raise ValueError(f"Excavator visual scale must be positive at {path}:{line_number}")
        if part in result:
            raise ValueError(f"Duplicate excavator part {part!r} in {path}")
        result[part] = MeshCalibration(numbers[:3], numbers[3:6], numbers[6])
    missing = {"base", "boom", "arm", "bucket"} - result.keys()
    if missing:
        raise ValueError(f"Missing excavator calibration parts in {path}: {sorted(missing)}")
    return result


def _calibration_frame(calibration: MeshCalibration) -> chrono.ChFramed:
    roll, pitch, yaw = (math.radians(value) for value in calibration.rotation_degrees)
    rotation = (
        chrono.QuatFromAngleZ(yaw)
        * chrono.QuatFromAngleY(pitch)
        * chrono.QuatFromAngleX(roll)
    )
    return chrono.ChFramed(vector(calibration.translation), rotation)


def _load_mesh(filename: str, scale: float) -> chrono.ChTriangleMeshConnected:
    path = ASSET_DIR / filename
    mesh = chrono.ChTriangleMeshConnected.CreateFromSTLFile(str(path))
    if mesh is None or mesh.GetNumVertices() == 0:
        raise RuntimeError(f"Could not load MathScavator9000 vehicle asset: {path}")
    mesh.Transform(chrono.VNULL, chrono.ChMatrix33d(scale))
    return mesh


def _set_imported_visual(
    body: chrono.ChBody,
    mesh: chrono.ChTriangleMeshConnected,
    name: str,
    frame: chrono.ChFramed,
    paint: tuple[float, float, float],
) -> None:
    body.GetVisualModel().Clear()
    shape = chrono.ChVisualShapeTriangleMesh(mesh, False)
    shape.SetName(name)
    shape.SetColor(color(paint))
    body.AddVisualShape(shape, frame)


def _style_cab(cabin: chrono.ChBody) -> None:
    glass = (0.08, 0.18, 0.22)
    add_box_visual(cabin, (0.035, 1.35, 0.72), (1.01, 0.0, 0.18), glass)
    add_box_visual(cabin, (1.20, 0.035, 0.72), (0.20, 0.91, 0.18), glass)
    add_box_visual(cabin, (1.20, 0.035, 0.72), (0.20, -0.91, 0.18), glass)
    add_box_visual(cabin, (2.15, 1.95, 0.12), (0.0, 0.0, 0.80), (0.92, 0.58, 0.02))


def _style_track(chassis: chrono.ChBody, center: tuple[float, float, float]) -> None:
    add_box_visual(chassis, (3.8, 0.55, 0.65), center, (0.12, 0.12, 0.12))
    for index in range(13):
        x = -1.72 + 3.44 * index / 12
        add_box_visual(
            chassis,
            (0.20, 0.60, 0.07),
            (center[0] + x, center[1], center[2] - 0.36),
            (0.20, 0.20, 0.19),
        )
        add_box_visual(
            chassis,
            (0.20, 0.60, 0.07),
            (center[0] + x, center[1], center[2] + 0.36),
            (0.20, 0.20, 0.19),
        )


def _style_track_rotor(rotor: chrono.ChBody) -> None:
    add_box_visual(rotor, (0.45, 0.50, 0.45), (0.0, 0.0, 0.0), (0.18, 0.18, 0.17))
    add_box_visual(rotor, (0.12, 0.58, 0.10), (0.0, 0.0, 0.25), (0.95, 0.65, 0.05))


def _add_joint_marker(body: chrono.ChBody, paint: tuple[float, float, float]) -> None:
    marker = chrono.ChVisualShapeSphere(0.16)
    marker.SetColor(color(paint))
    body.AddVisualShape(marker, chrono.ChFramed(chrono.ChVector3d(0.0, -0.42, 0.0)))


class ExcavatorModel:
    """Asset-backed excavator with imposed track drive and four motorized joints."""

    def __init__(
        self,
        system: chrono.ChSystem,
        show_rigid_ground: bool = True,
        show_calibration_markers: bool = False,
    ) -> None:
        self.system = system
        initial_body_count = len(system.GetBodies())
        self.visual_calibration_path = ASSET_DIR / "visual_calibration.cfg"
        calibration = _load_calibration(self.visual_calibration_path)

        self.ground = make_box(
            system,
            (30.0, 20.0, 0.1),
            (4.0, 0.0, -0.05),
            (0.38, 0.30, 0.18),
            visible=show_rigid_ground,
        )
        self.ground.SetFixed(True)
        self.chassis = make_box(system, (3.6, 2.4, 0.45), (0.0, 0.0, 0.75), (0.22, 0.22, 0.20))
        self.cabin = make_box(system, (2.0, 1.8, 1.5), (0.0, 0.0, 1.75), YELLOW)
        self.left_track = make_box(
            system, (0.20, 0.20, 0.20), (0.0, 1.35, 0.48), (0.12, 0.12, 0.12), visible=False
        )
        self.right_track = make_box(
            system, (0.20, 0.20, 0.20), (0.0, -1.35, 0.48), (0.12, 0.12, 0.12), visible=False
        )
        _style_track_rotor(self.left_track)
        _style_track_rotor(self.right_track)

        boom_pivot = (0.75, 0.0, 2.25)
        arm_pivot = (boom_pivot[0] + 6.24 * MATHSCAVATOR_SCALE, 0.0, boom_pivot[2])
        bucket_pivot = (
            arm_pivot[0] - 0.0091425 * MATHSCAVATOR_SCALE,
            0.0,
            arm_pivot[2] - 2.9796 * MATHSCAVATOR_SCALE,
        )
        self.boom = make_box(system, (2.8, 0.35, 0.35), boom_pivot, YELLOW)
        self.arm = make_box(system, (2.2, 0.30, 0.30), arm_pivot, YELLOW)
        self.bucket = make_box(system, (0.9, 0.52, 1.0), bucket_pivot, (0.90, 0.55, 0.03))
        _style_cab(self.cabin)

        meshes = {
            "base": _load_mesh("base_link.STL", calibration["base"].scale),
            "boom": _load_mesh("chassis_boom_link.STL", calibration["boom"].scale),
            "arm": _load_mesh("boom_stick_link.STL", calibration["arm"].scale),
            "bucket": _load_mesh("stick_bucket_link.STL", calibration["bucket"].scale),
        }
        self._meshes = meshes
        self.imported_vertex_count = sum(mesh.GetNumVertices() for mesh in meshes.values())
        _set_imported_visual(
            self.chassis,
            meshes["base"],
            "MathScavator9000 undercarriage",
            _calibration_frame(calibration["base"]),
            (0.20, 0.20, 0.20),
        )
        _set_imported_visual(
            self.boom,
            meshes["boom"],
            "MathScavator9000 boom",
            _calibration_frame(calibration["boom"]),
            YELLOW,
        )
        _set_imported_visual(
            self.arm,
            meshes["arm"],
            "MathScavator9000 stick",
            _calibration_frame(calibration["arm"]),
            YELLOW,
        )
        _set_imported_visual(
            self.bucket,
            meshes["bucket"],
            "MathScavator9000 bucket",
            _calibration_frame(calibration["bucket"]),
            (0.90, 0.55, 0.03),
        )
        add_box_collision(self.bucket, (0.88, 0.52, 1.00), (-0.384, -0.002, 0.044))
        if show_calibration_markers:
            _add_joint_marker(self.boom, (1.0, 0.0, 0.8))
            _add_joint_marker(self.arm, (0.1, 1.0, 0.1))
            _add_joint_marker(self.bucket, (0.0, 0.9, 1.0))

        left_track_center = (0.0, 1.35, -0.27)
        right_track_center = (0.0, -1.35, -0.27)
        _style_track(self.chassis, left_track_center)
        _style_track(self.chassis, right_track_center)
        add_box_collision(self.chassis, (3.8, 0.55, 0.72), left_track_center)
        add_box_collision(self.chassis, (3.8, 0.55, 0.72), right_track_center)

        self.drive = DifferentialDrive(system, self.chassis, self.ground, 2.7)
        hinge_y = chrono.QuatFromAngleX(chrono.CH_PI_2)
        self.joints = [
            make_angle_motor(system, self.cabin, self.chassis, (0.0, 0.0, 1.0)),
            make_angle_motor(system, self.boom, self.cabin, boom_pivot, hinge_y),
            make_angle_motor(system, self.arm, self.boom, arm_pivot, hinge_y),
            make_angle_motor(system, self.bucket, self.arm, bucket_pivot, hinge_y),
        ]

        self.left_track_drive = chrono.ChLinkMotorRotationSpeed()
        self.left_track_drive.Initialize(
            self.left_track,
            self.chassis,
            chrono.ChFramed(chrono.ChVector3d(0.0, 1.35, 0.48), hinge_y),
        )
        self.left_track_drive.SetSpeedFunction(chrono.ChFunctionConst(0.0))
        system.Add(self.left_track_drive)
        self.right_track_drive = chrono.ChLinkMotorRotationSpeed()
        self.right_track_drive.Initialize(
            self.right_track,
            self.chassis,
            chrono.ChFramed(chrono.ChVector3d(0.0, -1.35, 0.48), hinge_y),
        )
        self.right_track_drive.SetSpeedFunction(chrono.ChFunctionConst(0.0))
        system.Add(self.right_track_drive)
        self.body_count = len(system.GetBodies()) - initial_body_count

    def set_drive_speeds(self, left: float, right: float) -> None:
        self.drive.set_speeds(left, right)
        self.left_track_drive.SetSpeedFunction(chrono.ChFunctionConst(left))
        self.right_track_drive.SetSpeedFunction(chrono.ChFunctionConst(right))

    def set_articulation_targets(self, swing: float, boom: float, arm: float, bucket: float) -> None:
        for motor, target in zip(self.joints, (swing, boom, arm, bucket), strict=True):
            motor.SetAngleFunction(chrono.ChFunctionConst(target))

    def advance(self, step: float) -> None:
        self.drive.synchronize(self.system.GetChTime() + step)
        self.system.DoStepDynamics(step)

    def get_chassis_position(self) -> chrono.ChVector3d:
        position = self.chassis.GetPos()
        return chrono.ChVector3d(position.x, position.y, position.z)

    def get_chassis_heading(self) -> float:
        return self.chassis.GetRot().GetCardanAnglesXYZ().z

    def get_joint_angles(self) -> tuple[float, float, float, float]:
        return tuple(motor.GetMotorAngle() for motor in self.joints)

    def get_track_angles(self) -> tuple[float, float]:
        return self.left_track_drive.GetMotorAngle(), self.right_track_drive.GetMotorAngle()

    def get_articulation_body_positions(self) -> tuple[chrono.ChVector3d, ...]:
        return tuple(
            chrono.ChVector3d(position.x, position.y, position.z)
            for position in (self.boom.GetPos(), self.arm.GetPos(), self.bucket.GetPos())
        )
